ironfish-cli/src/commands/accounts/repair.ts

```diff
@@ -0,0 +1,197 @@
+export default class Repair extends IronfishCommand {
+  static hidden = false
+
+  static description = `Repairs wallet database stores`
+
+  static flags = {
+    ...LocalFlags,
+  }
+
+  static args = [
+    {
+      name: 'account',
+      parse: (input: string): Promise<string> => Promise.resolve(input.trim()),
+      required: false,
+      description: 'Name of the account to repair the database for',
+    },
+  ]
+
+  async start(): Promise<void> {
+    const { args } = await this.parse(Repair)
+
+    const node = await this.sdk.node()
+    await NodeUtils.waitForOpen(node)
+
+    const account = this.loadAccount(node.wallet, args.account)
+
+    this.log(`Repairing wallet for account ${account.name}`)
+
+    this.log('Repairing balance')
+    await this.repairBalance(account, node.wallet.walletDb, node.chain)
+
+    this.log('Repairing nullifierToNote')
+    await this.repairNullifierToNoteHash(account, node.wallet.walletDb)
+
+    this.log('Repairing sequenceToNoteHash')
+    await this.repairSequenceToNoteHash(account, node.wallet.walletDb)
+  }
+
+  private loadAccount(wallet: Wallet, accountName: string | undefined): Account {
+    if (accountName) {
+      const account = wallet.getAccountByName(accountName)
+      if (account) {
+        return account
+      }
+      throw new Error(`No account with name ${accountName}`)
+    }
+
+    const defaultAccount = wallet.getDefaultAccount()
+    if (defaultAccount) {
+      return defaultAccount
+    }
+
+    throw new Error('Could not find an account to repair.')
+  }
+
+  private async repairBalance(
+    account: Account,
+    walletDb: WalletDB,
+    chain: Blockchain,
+  ): Promise<void> {
+    let unconfirmedBalance = 0n
+
+    let noteUnspentMismatches = 0
+
+    await walletDb.db.transaction(async (tx) => {
+      for await (const decryptedNoteValue of account.getNotes()) {
+        const transactionValue = await account.getTransaction(
+          decryptedNoteValue.transactionHash,
+          tx,
+        )
+
+        Assert.isNotUndefined(
+          transactionValue,
+          `Account has a note but is missing the transaction that it received the note from. ${RESCAN_MESSAGE}`,
+        )
+
+        if (!decryptedNoteValue.nullifier) {
+          if (transactionValue.sequence) {
+            throw new Error(
+              `Transaction marked as on chain, but note missing nullifier. ${RESCAN_MESSAGE}`,
+            )
+          }
+
+          continue
+        }
+
+        const spent = await chain.nullifiers.contains(decryptedNoteValue.nullifier)
+
+        if (spent && !decryptedNoteValue.spent) {
+          noteUnspentMismatches++
+
+          await walletDb.saveDecryptedNote(
+            account,
+            decryptedNoteValue.hash,
+            {
+              ...decryptedNoteValue,
+              spent,
+            },
+            tx,
+          )
+        } else if (!spent) {
+          unconfirmedBalance += decryptedNoteValue.note.value()
```

Comment:
We may also need to check whether a transaction is already expired. I've identified an instance where a wallet seems to have re-synced a transaction that was already expired.